File: AkitaModel/PaymentBase.cs
using System;
using System.Text.Json.Serialization;

namespace Skyware.Lis.AkitaModel;

/// <summary>
/// Base (abstract) class for payments.
/// </summary>
[JsonDerivedType(typeof(CashPayment))]
[JsonDerivedType(typeof(CardPayment))]
[JsonDerivedType(typeof(VoucherPayment))]
public abstract class PaymentBase
{

    /// <summary>
    /// PK of the item in LIS iLab.
    /// </summary>
    public int Id { get; set; }

    /// <summary>
    /// Date/time the payment was made.
    /// </summary>
    public DateTime Date { get; set; }

    /// <summary>
    /// Payment method.
    /// </summary>
    public abstract PaymentMethod Method { get; }

    /// <summary>
    /// Amount of the payment.
    /// </summary>
    public decimal Amount { get; set; }

}

File: AkitaMotelTS/src/PaymentBase.ts
/**
 * Skyware.Lis.AkitaModel
 */

import { PaymentMethod } from './PaymentMethod';

/**
 * Base (abstract) class for payments.
 */
export abstract class PaymentBase {
  /**
   * PK of the item in LIS iLab.
   */
  id!: number;

  /**
   * Date/time the payment was made.
   */
  date!: Date;

  /**
   * Payment method.
   */
  abstract get method(): PaymentMethod;

  /**
   * Amount of the payment.
   */
  amount!: number;
}
